filebeat/input/file/states.go

```diff
@@ -0,0 +1,157 @@
+package file
+
+import (
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+// States handles list of FileState. One must use NewStates to instantiate a
+// file states regisry. Using the zero-value is not safe.
+type States struct {
+	sync.RWMutex
+
+	// states store
+	states []State
+
+	// idx maps state IDs to state indexes for fast lookup and modifications.
+	idx map[string]int
```

Comment:
What is the reason you changed your mind to the separate map string for keeping track of the ids?

Reply:
Too many copies :)

The states mapping has these very frequent access patterns:
- access single entry by ID
- loop/copy all states

Performance wise, both access patterns are equally important, as they are run very very often (basically per event).

Optimizing the access by ID case (using map[string]state) I found the profiling being dominated by extra copies, even if only one field is required. Introducing an index on array indexes, we still have the well working 'all-access loop' + we optimize single entry access to be about O(1) complexity.

Kind of best of both worlds with a little constant overhead when actually removing entries from the registry. As state update/add + iteration + copy are the most often used operations, the added overhead on removal is pretty much worth it.